### PR TITLE
CTP 2496 - now using file API for news images

### PR DESCRIPTION
**Pull request by opitz**
plus cleanup for GHA tests.

### CODE REVIEW

**Comment by opitz** [comment]
@aspark21 Deployed on 43-codebase now.

File: lib.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    // Because the images are uploaded in settings they are stored with system context (id = 1).
+    // To retrieve them from inside a block we have to force the system context here.
+    $contextid = 1;
+    $file = $fs->get_file($contextid, 'block_news', $filearea, $itemid, $filepath, $filename);
+    if (!$file) {
+        // The file does not exist.
+        return false;
+    }
+
+    // We can now send the file back to the browser.
```
Comment (stuartlamour) [inline lib.php:80]: These magic numbers look interesting - can we comment what they are @opitz ?

File: settings.php [hunk 2]
```diff
@@ -57,14 +58,14 @@
...
+                'block_news/image'.$i,
                 get_string('image', 'block_news'),
                 get_string('image_help', 'block_news'),
-                $default,
-                PARAM_RAW
-            );
-            $settings->add($setting);
+                'block_news',
+                $i
+            ));
```
Comment (stuartlamour) [inline settings.php:68]: Seems to be missing $settings->add($setting);

Comment (opitz) [reply settings.php:68]: No need to as I add it directly to $settings.

File: lib.php (hunk 1 above)
Comment (opitz) [reply lib.php]: Done.

File: settings.php [hunk 3]
```diff
@@ -57,14 +58,14 @@
...
-            $setting = new admin_setting_configtext('block_news/image'.$i,
+            // Adding a file upload element for an image.
+            $settings->add(new admin_setting_configstoredfile(
+                'block_news/image'.$i,
                 get_string('image', 'block_news'),
                 get_string('image_help', 'block_news'),
-                $default,
-                PARAM_RAW
-            );
-            $settings->add($setting);
```
Comment (stuartlamour) [inline settings.php:67]: Having this single setting use: 

$settings->add(all the things);

while the other settings all use:

$setting = thing;
$settings->add($setting) 

feels inconsistant.

**Review by stuartlamour — APPROVED** [review]

**Review by stuartlamour — APPROVED** [review]

**Comment by aspark21** [comment]
ok, so how do we handle the migration for this? Do files need to be re-uploaded when we deploy this updated version?

If so, can you document steps into the next deployment https://wiki.ucl.ac.uk/display/ISMoodle/End+of+DLE+Sprint+1.2+23-24+release

**Comment by opitz** [comment]
> ok, so how do we handle the migration for this? Do files need to be re-uploaded when we deploy this updated version?
> 
> If so, can you document steps into the next deployment https://wiki.ucl.ac.uk/display/ISMoodle/End+of+DLE+Sprint+1.2+23-24+release

@aspark21 @stuartlamour I thought it to be a nice exercise to perform an upgrade task on this - so I added one. 
It will take the image file from the currently used URL, store it into the Moodle file area and then use that.

Please review again...!

File: db/upgrade.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                $fs = get_file_storage();
+                $file = $fs->get_file($contextid, $component, $filearea, $itemid, $filepath, $filename);
+                if (!$file) {
+                    // Only add the file if it does not already exist. Should not be the case, but check anyway.
+                    $fs->create_file_from_url($fileinfo, $image);
+                }
+                set_config( 'image'.$i, $filepath.$filename, 'block_news');
+            }
+        }
+        upgrade_block_savepoint(true, 2023111000, 'news');
```
Comment (aspark21) [inline db/upgrade.php:62]: I think you're missing the plugin type in that array

e.g. https://github.com/ucl-isd/moodle-local_courserollover/blob/main/db/upgrade.php#L108

Comment (aspark21) [reply db/upgrade.php:62]: scratch that, I see it's 'upgrade_block_savepoint' rather than 'upgrade_plugin_savepoint'

**Comment by aspark21** [comment]
@opitz ok upgrade step looks good to me. Have tested it.

The only lingering concern is that this works well with the objectfs / S3 filestore. Our Preview setup is read-only but maybe you can test on your local machine to add the file and get it to upload to S3 and be served that way.
There's some weird gunk at the end of the filename which I'm slightly worried may not work if offloaded to S3
`pluginfile.php/6294641/block_news/block_news/3/s_uk_ucl_61_web_ready800x500_0.jpg%3Fitok%3D70kFDOd-`

**Comment by aspark21** [comment]
the itok parameter at the end is a drupla thing so can remove that from the url before the upgrade

**Review by aspark21 — APPROVED** [review]

**Comment by opitz** [comment]
> @opitz ok upgrade step looks good to me. Have tested it.
> 
> The only lingering concern is that this works well with the objectfs / S3 filestore. Our Preview setup is read-only but maybe you can test on your local machine to add the file and get it to upload to S3 and be served that way. There's some weird gunk at the end of the filename which I'm slightly worried may not work if offloaded to S3 `pluginfile.php/6294641/block_news/block_news/3/s_uk_ucl_61_web_ready800x500_0.jpg%3Fitok%3D70kFDOd-`

I was thinking to replace any image name from the URL with a generic one (e.g. news_image_1) as this is only used in the files table and would be replaced with a real file name once it will be changed in the settings. Or I could clip everything after the 1st “%”?

**Pull request merged**